apps/dashboard/app/controllers/settings_controller.rb

```diff
@@ -1,17 +1,45 @@
 # frozen_string_literal: true
 
 # The Controller for user level settings /dashboard/settings.
+# Current supported settings: profile, announcement
 class SettingsController < ApplicationController
   include UserSettingStore
-  ALLOWED_SETTINGS = [:profile].freeze
+  ALLOWED_SETTINGS = [:profile, :announcement].freeze
 
   def update
     new_settings = read_settings(settings_param)
-    update_user_settings(new_settings) unless new_settings.empty?
+    if new_settings.empty?
+      respond_to do |format|
+        format.html { redirect_to root_url, alert: I18n.t('dashboard.settings_invalid_request') }
+        format.json { head :bad_request }
+      end
 
-    logger.info "settings: updated user settings to: #{new_settings}"
+      return
+    end
+
+    # Only ALLOWED_SETTINGS methods will be called
+    setting, value = new_settings.first
+    send("update_#{setting}", value)
```

Comment:
Yea I feel like this is a bit like `working by coincidence`. This is the first PR to update the settings controller, so we're just at that spot, but I feel like it should be easier. 

That is, I don't think we need different methods for different things we're updating. Or at least I hope not just because that'll explode. 

Can't we follow what we had before? I guess I'm really asking if we need all these special methods or if we can't just merge the 2 hashes and save?

* filter the parameters coming into the controller
* read the current settings
* merge
* save

Reply:
I think we are better defining the contract of settings as we add more. The new contract means that you can only update one setting at the time using the `SettingsController`. And the specific methods is to be able to add more logic into each setting into the backend. Otherwise it will have to move somewhere else.

My first implementation was as you suggested, but I had to allow to submit a hash as a value to support multiple announcements. This is the pseudocode for the that solution:

```ruby
SettingsController
...
  ALLOWED_SETTINGS = [:profile, { announcements: {} }].freeze
...
  def read_settings(params)
    {}.tap do |settings|
      params&.each do |key, value|
        value = value.to_h if value.is_a?(ActionController::Parameters)
        settings[key] = value
      end
    end
  end
```

```erb
# The button to summit the setting
<%=
    button_to(
      settings_path("settings[announcements][#{announcement.id}]" => '2024-07-18 12:00:00'),
      method: :post,
      class: %w[btn btn-warning me-md-2],
    ) do
      'Accept'
    end
  %>
```

Reply:
Actually, now that I look at the code, this could be simplified to:

```ruby
  def read_settings(params)
    params.to_h
  end
```

My only concern is now allowing a hash as a parameter. Do you see this as an issue?

Reply:
I have updated the SettingsController as per your suggestion. Looking at it now, it is not as complicated as my first implementation. Let me know what you think

Reply:
Yea I get all that, that it'll get complicated as we go. IDK about passing a hash into from a `form`, so I guess we'll have to work that out. If it's simple now and works, then I'm happy and we can sort of continue to refine it.

That said - at some point we're likely to make a page so that folks can edit their settings, so it's likely one giant `form` with all possible configurations.  So this is kinda why I'm pushing for simplicity, so that we can provide an `edit` page someday.